r/tests/testthat/test-dplyr-summarize.R

```diff
@@ -227,6 +228,19 @@ test_that("Group by n_distinct() on dataset", {
       collect(),
     tbl
   )
+  # Without groupby
+  expect_dplyr_equal(
+    input %>%
+      summarize(distinct = n_distinct(lgl, na.rm = FALSE)) %>%
+      collect(),
+    tbl
+  )
+  expect_dplyr_equal(
+    input %>%
+      summarize(distinct = n_distinct(lgl, na.rm = TRUE)) %>%
```

Comment:
What about more than one column? What about 0 columns? Both work in on data.frames in R:

```
> mtcars %>% summarize(n_distinct())
  n_distinct()
1            0
> mtcars %>% summarize(n_distinct(hp, cyl))
  n_distinct(hp, cyl)
1                  23
```

Reply:
Good catch @nealrichardson , thanks!
Let me try to address this. I guess the kernel will remain as unary in C++, but I need to change the definitions in `dplyr-functions.R`  for `agg_funcs$n_distinct` to support that behavior.
@ianmcook Let me know if you have more ideas here, I could use some for guidance :)

Reply:
I think we will need to save this for a follow-up.

As @aucahuasi says, the `count_distinct` kernel is unary. In theory we could make it count the distinct combinations of values in two or more columns by sticking all the columns into a struct and passing the struct to `count_distinct`. So for example we would replace
```r
summarise(n_distinct(x, y))
```
with
```r
summarise(n_distinct(arrow_make_struct(x, y, options = list(field_names = c("x", "y")))))
```

But there are two challenges there:

1. This would require doing more hacky recurse-through-the-AST-and-replace-stuff, like in #11018
2. I don't think this `count_distinct` kernel currently accepts nested types.

I opened ARROW-14209 for the follow-up.

Reply:
That's correct: It doesn't accept nested types. That's another follow-up too (not only for this kernel, but for others count distinct related ones as well)